File: my-folder/1408-find-the-smallest-divisor-given-a-threshold/solution.cpp
class Solution {
public:
    int check(vector<int> v,int divisor,int threshold){
        int sum = 0;
        for(int i=0;i<v.size();i++){
            sum += ceil((double)(v[i]) / (double)(divisor));
            if(sum>threshold) return 0;
        }
        // if(sum<=threshold) return 1;
        // else return 0;
        return 1;
    }
    int smallestDivisor(vector<int>& nums, int threshold) {
        int n = nums.size();
        if (n > threshold) return -1;
        int low = 1;
        int high = *max_element(nums.begin(),nums.end());
        while(low<=high){
            int mid = low + (high-low)/2;
            int canIt = check(nums,mid,threshold);
            if(canIt==1) high = mid-1;
            else low = mid+1;

        }
        return low;
        
    }
};

File: my-folder/1408-find-the-smallest-divisor-given-a-threshold/solution.java
class Solution {
    boolean solve(int[] nums,int threshold,int div){
        int sum = 0;
        for(int i=0;i<nums.length;i++){
            sum+=(nums[i]+div-1)/div;
        }
        return sum<=threshold;
    }
    public int smallestDivisor(int[] nums, int threshold) {
        int ans = Integer.MAX_VALUE;
        int right = Arrays.stream(nums).max().getAsInt();  
        int left = 1;
        while(left<=right){
            int mid = (left+right)/2;
            if(solve(nums,threshold,mid)){
                ans = mid;
                right = mid-1;
            }
            else left = mid+1;
        }
        return ans;
    }
}

